pwnlib/libcdb.py

```diff
@@ -25,6 +26,52 @@
 
 HASHES = ['build_id', 'sha1', 'sha256', 'md5']
 
+# https://gitlab.com/libcdb/libcdb wasn't updated after 2019,
+# but still is a massive database of older libc binaries.
+def provider_libcdb(hex_encoded_id, hash_type):
+    # Build the URL using the requested hash type
+    url_base = "https://gitlab.com/libcdb/libcdb/raw/master/hashes/%s/" % hash_type
+    url      = urllib.parse.urljoin(url_base, hex_encoded_id)
+
+    data     = b""
+    while not data.startswith(b'\x7fELF'):
+        log.debug("Downloading data from LibcDB: %s", url)
+        data = wget(url, timeout=20)
+
+        if not data:
+            log.warn_once("Could not fetch libc for %s %s from libcdb", hash_type, hex_encoded_id)
+            break
+        
+        # GitLab serves up symlinks with
+        if data.startswith(b'..'):
+            url = os.path.dirname(url) + '/'
+            url = urllib.parse.urljoin(url.encode('utf-8'), data)
+    return data
+
+# https://libc.rip/
+def provider_libc_rip(hex_encoded_id, hash_type):
+    # Build the request for the hash type
+    # https://github.com/niklasb/libc-database/blob/master/searchengine/api.yml
+    if hash_type == 'build_id':
+        hash_type = 'buildid'
+    url    = "https://libc.rip/api/find"
+    params = {hash_type: hex_encoded_id}
+
+    result = requests.post(url, json=params, timeout=20)
+    if result.status_code != 200 or len(result.json()) == 0:
+        log.warn_once("Could not find libc for %s %s on libc.rip", hash_type, hex_encoded_id)
+        log.debug("Error: %s", result.text)
+        return None
+
+    libc_match = result.json()
+    assert len(libc_match) == 1, 'Invalid libc.rip response.'
+
+    url = libc_match[0]['download_url']
+    log.debug("Downloading data from libc.rip: %s", url)
+    return wget(url, timeout=20)
```

Comment:
What happens when this fails?  The libcdb providers need to fail gracefully -- e.g. the original just threw a warning on `if not data`

Reply:
Right now this relies on the loop calling the different providers. If none of the providers worked, it logs a warning. I can add a more explicit check here too if you'd like?

Reply:
The concern is that `wget` may throw an exception that isn't caught.

It looks like the original code didn't handle this well, either.

```
>>> wget('http://example-does-not-exist.tld')                                                                                                                                                                                                 
ConnectionError: HTTPConnectionPool(host='example-does-not-exist.tld', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x102d03910>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
```

Reply:
Ah, I'll catch those.